docs: add decision record for naming conventions (#2067)
eclipse-dataspaceconnector/DataSpaceConnector: docs/developer/decision-records/2022-10-10-naming-conventions/README.md

and [federated catalog](https://github.com/eclipse-dataspaceconnector/FederatedCatalog))
are published with the same version under this same _groupId_.

The _artifactId_ is the name of the jar without a version. In the EDC project, the _artifactId_ is identical
to the _moduleId_. For example:

```kotlin
publishing {
    publications {
        create<MavenPublication>("data-plane-framework") {
            artifactId = "data-plane-framework"
            from(components["java"])
        }
    }
}
```

For versioning, see [here](../2022-07-06-release-automation/).

#### Java Package

Following Oracle's [Java package naming conventions](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html),
package names are written in all lower case. In addition, package names use the reversed domain name,
for example, `com.example.mypackage` for a package named `mypackage` created by a programmer at `example.com`.

Accordingly, in the EDC project, aligned with the _groupId_, Java packages start with `org.eclipse.edc.*`.
Corresponding to the _moduleId_, this is followed by a unique path. To remain with our example, the
class `DataPlaneFrameworkExtension` is located at:

```java

> package org.eclipse.dataspaceconnector.dataplane.framework;

Reviewer: update `dataspaceconnector` --> `edc`

Author: done

```

> In this context, everything after `org.eclipse.edc` is sorted first by field/discipline and then by scope.

Reviewer: Instead of "field/discipline", can we say "component"?

Author: reformulated

For example, a policy spi will not be located at `org.eclipse.edc.spi.policy` but at `org.eclipse.edc.policy.spi`.

#### Configuration Property

Following the [policy scopes](../2022-03-15-policy-scopes/), configuration properties are using a hierarchical
dot notation. They do **not** start with an `edc` or similar Eclipse related prefix.

> Configuration properties should fit to related _moduleIds_ or package names and be grouped by scope.

Reviewer: I would prefer the config to follow module ids, so they are decoupled from packages, as we don't want to introduce configuration migration issues if package refactoring is done in the future. If we move a a class across a module this may happen but it is less likely than a package.

Reviewer: Also, can we include some examples here?

Author: done, reformulated

Our goals by introducing naming conventions are:
- Easy navigation of either Gradle modules, Maven artifacts or the Java package structure.
- Unique naming of config properties (especially with regard to custom properties introduced by extensions).
- Accordance to the Eclipse Foundation's guidelines and release processes.
- Elimination of split packages.

## Approach

### Gradle module name

Gradle modules must have unique names across the entire build classpath, even if they are located in
different module paths. This is because of a bug in Gradle itself where Gradle will erroneously report
a "cyclic dependency" if this rule is violated. The following hypothetical example would constitute
such a violation:

```kotlin
// settings.gradle.kts
include(":core:common:transfer-process")
include(":extensions:sql:transfer-process")
```

The EDC project has checks in place to make sure module IDs are unique.

> Rule 1: Modules must have unique names.

In addition, the _module name_ should give a hint what is in the module, without being too verbose. The
earlier example would be a bad one, because "transfer-process" does not indicate what the contents could

> be. This is especially important because Maven's _artifactId_ must be equal to module names.

Reviewer 2: ```suggestion
be. This is especially important because we require the Maven's _artifactId_ to be equal to module names.
```
That makes it clearer that it is not actually a requirement imposed by Maven or Gradle

Here are some bad examples:
- `:core:common:transfer-process:validation`: bad because "validation" is likely to be not unique and isolated it only indicates that it has to do with validation, but not in _what context_.
- `:core:dataplane:framework`: again, "framework" is liable to cause conflicts, and in addition, it's a very generic, unspecific term

Refactoring these bad examples, we could arrive at these:
- `:core:common:transfer-process:transfer-process-validation`: could contain validation functions
- `:core:dataplane:dataplane-framework`: would contain default impls and platform code for the dataplane

> Rule 2: Module names should indicate what the contents are.

> > Rule 3: Module names must be identical to the Maven artifactId (published one).

Reviewer 2: ```suggestion
> Rule 3: Module names must be identical to the Maven artifactId (if one is published).
```
not all modules publish an artifact

### Maven artifactId

The EDC project uses the same `groupId = "org.eclipse.edc"` across all sub-projects, which means all
_artifactIds_ must be unique across multiple software components to avoid conflicts.

> Rule 4: A Maven artifactId must be unique within the groupId.

> Rule 5: A Maven artifactId must be identical to the module name (cf. [Rule 3](#gradle-module-name)).

### Java package name

Following Oracle's [Java package naming conventions](https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html),
EDC's base package name is `org.eclipse.edc`, followed by domain and then function. For example, the
Policy SPI should be located at `org.eclipse.edc.policy.spi` rather than `org.eclipse.edc.spi.policy`.
Here, `policy` is the "domain", i.e. the thematic context, and `spi` would be the "function", i.e. what

> kind of package it is or: what purpose it serves.

Reviewer 2: ```suggestion
kind of package it is or what purpose it serves.
```

The module name could be a helpful reference for the package name, replacing dashes with dots.

Other positive examples would be:
- `org.eclipse.edc.transferprocess.validation`: domain is `transferprocess`, and it's a package that deals with validation
- `org.eclipse.edc.dataplane.framework.manager`: here the domain is `dataplane.framework`, so all code should be beneath that directory.

This helps to avoid split packages.

> Rule 6: Package names should first contain the domain, and then the function.

### Configuration properties

> Configuration properties should have a unique prefix `edc` to avoid clashes, when EDC gets embedded

Reviewer 3: I remember we decided to remove the `edc` prefix as it was unnecessary, using the module name as prefix instead. I am correct?

Author: Fine with removing, I had a look, I noted it down as decision. With the module name, I am not sure.. this could get ugly :D However, this would be simple and self-descriptive

Reviewer 3: I have no strong opinion on this, I just remember we had a different decision.
Just for confirm: @paullatzelsperger @jimmarino @ronjaquensel are we ok with the `edc` prefix for config settings? Wasn't it considered implicit?

Reviewer 4: @ndr-brt I also think we decided to omit this prefix, but thinking that the EDC may be embedded in/combined with other frameworks where potentially configuration might get mixed, I think it makes sense to add the prefix. But I also don't have a strong opinion on this.

Author: if nobody has a strong optinion, I would merge this as-is
